### PR TITLE
Constant propagation of with2t expression

### PR DESCRIPTION
**Pull request by XLiZHI**

### CODE REVIEW

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
@XLiZHI: can you run this PR over the SV-COMP benchmarks?

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -141,7 +149,7 @@ bool goto_symex_statet::constant_propagation(const expr2tc &expr) const
     bool noconst = true;
 
     expr->foreach_operand([this, &noconst](const expr2tc &e) {
-      if (noconst && !constant_propagation(e))
+      if (noconst && (!constant_propagation(e) || is_with2t(e)))
```
Comment (mikhailramalho) [inline src/goto-symex/goto_symex_state.cpp]: is this needed?

File: src/util/simplify_expr2.cpp [hunk 2]
```diff
@@ -890,6 +890,26 @@ expr2tc member2t::do_simplify() const
...
+  {
+    auto current_source = source_value;
+
+    // Traverse through nested 'with' expressions
+    while (is_with2t(current_source))
+    {
+      const auto &with = to_with2t(current_source);
+
+      // Check if the member matches the update field
+      if (member == to_constant_string2t(with.update_field).value)
```
Comment (mikhailramalho) [inline src/util/simplify_expr2.cpp]: is `with.update_field` always a constant_string? If it isn't, you need to check it before calling `to_constant_string`

**Review by mikhailramalho — COMMENTED** [review]
Please check if this doesn't make performance worse, we saw regressions before when enabling constant folding of with expressions

File: src/goto-symex/goto_symex_state.cpp (hunk 1 above)
Comment (XLiZHI) [reply src/goto-symex/goto_symex_state.cpp]: Yes, It is necessary and without it there will be serious performance issues.

File: src/util/simplify_expr2.cpp (hunk 2 above)
Comment (XLiZHI) [reply src/util/simplify_expr2.cpp]: It is contant_string.

File: regression/esbmc/github_690_z3_cast_success/test.desc [hunk 3]
```diff
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.c
---ir --z3
+--ir --z3 -Wno-error=implicit-function-declaration
```
Comment (mikhailramalho) [inline regression/esbmc/github_690_z3_cast_success/test.desc:3]: This is an unrelated change

File: regression/esbmc-cpp/cpp/ch16_9/test.desc [hunk 4]
```diff
@@ -1,4 +1,4 @@
 KNOWNBUG
 main.cpp
---unwind 10 --no-unwinding-assertions --timeout 900
+--unwind 2 --no-unwinding-assertions --timeout 900
```
Comment (mikhailramalho) [inline regression/esbmc-cpp/cpp/ch16_9/test.desc]: is this related to this patch?

File: regression/esbmc/github_2119/main.c [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+typedef struct buffer buffer;
+
+int main () {
+  int counter = 6;
+  buff.size = 6;
+  buff.limit = 3;
+
+  for (int i = 0; i < buff.size; i++) {
+    if(buff.limit == i && !buff.counter[i])
+      counter--;
```
Comment (mikhailramalho) [inline regression/esbmc/github_2119/main.c]: whitespace

File: src/goto-symex/goto_symex_state.cpp (hunk 1 above)
Comment (mikhailramalho) [reply src/goto-symex/goto_symex_state.cpp]: Then at least add a comment explaining why this is here. Can you share an example where this is required?

File: src/util/simplify_expr2.cpp (hunk 2 above)
Comment (mikhailramalho) [reply src/util/simplify_expr2.cpp]: We don't have this constraint written anywhere, `update_field` is declared as an `expr2tc`, so let's add a check here.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
The new test case you add works just fine for me without your patch: 
```
$ esbmc main.c
ESBMC version 7.7.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing main.c
Converting
Generating GOTO Program
GOTO program creation time: 0.398s
GOTO program processing time: 0.001s
Starting Bounded Model Checking
Unwinding loop 2 iteration 1   file main.c line 17 column 3 function main
Unwinding loop 2 iteration 2   file main.c line 17 column 3 function main
Unwinding loop 2 iteration 3   file main.c line 17 column 3 function main
Unwinding loop 2 iteration 4   file main.c line 17 column 3 function main
Unwinding loop 2 iteration 5   file main.c line 17 column 3 function main
Unwinding loop 2 iteration 6   file main.c line 17 column 3 function main
Symex completed in: 0.007s (22 assignments)
Slicing time: 0.000s (removed 21 assignments)
Generated 13 VCC(s), 0 remaining after simplification (1 assignments)
BMC program time: 0.007s

VERIFICATION SUCCESSFUL
```
Please add a test case where we can see this patch in action

File: regression/esbmc/github_690_z3_cast_success/test.desc (hunk 3 above)
Comment (XLiZHI) [reply regression/esbmc/github_690_z3_cast_success/test.desc:3]: This is relevant. It passed in llvm13 in this PR, but llvm16 has stricter checks and therefore needs to add this flag.

Comment (intrigus-lgtm) [reply regression/esbmc/github_690_z3_cast_success/test.desc:3]: Just curious, shouldn't this have been caught by CI when updating to llvm16?
Why only in this PR?

**Comment by lucasccordeiro** [comment]
@XLiZHI : Our benchexec is now running again. Could you please run this PR over the SV-COMP benchmarks?

File: src/goto-symex/goto_symex_state.cpp (hunk 1 above)
Comment (mikhailramalho) [reply src/goto-symex/goto_symex_state.cpp]: Use the comment to describe why this is needed. It is obvious that we skip `with2t`s but why are we doing it?

File: regression/esbmc/github_690_z3_cast_success/test.desc (hunk 3 above)
Comment (mikhailramalho) [reply regression/esbmc/github_690_z3_cast_success/test.desc:3]: IIRC, we ignore the results if it's marked as a KNOWBUG

**Comment by lucasccordeiro** [comment]
This PR (https://github.com/esbmc/esbmc/actions/runs/11776848285):

````
Statistics:          30300 Files
  correct:           16446
    correct true:    11011
    correct false:    5435
  incorrect:            28
    incorrect true:     18
    incorrect false:    10
  unknown:           13826
  Score:             26721 (max: 51772)
````

Master ( https://github.com/esbmc/esbmc/actions/runs/11529888639):

````
Statistics:          30300 Files
  correct:           16532
    correct true:    11085
    correct false:    5447
  incorrect:            28
    incorrect true:     18
    incorrect false:    10
  unknown:           13740
  Score:             26881 (max: 51772)
````

@XLiZHI: could you please check the `unknown` cases to confirm whether they are crashes or timeouts?

**Review by mikhailramalho — APPROVED** [review]
LGTM, let's just update the comment.

**Comment by lucasccordeiro** [comment]
@mikhailramalho: do you refer to this comment?

image

**Comment by mikhailramalho** [comment]
No, this one:

image

It tells me nothing about why we are only focusing on structs. I would expect a comment saying that enabling others will regress performance and hopefully a test case where this regression is reproducible.

**Comment by XLiZHI** [comment]
> It tells me nothing about why we are only focusing on structs. I would expect a comment saying that enabling others will regress performance and hopefully a test case where this regression is reproducible.

I have removed this change.

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @XLiZHI.
